src/System/Taffybar/Widgets/PollingBar.hs

```diff
@@ -7,27 +7,27 @@ module System.Taffybar.Widgets.PollingBar (
   BarDirection(..),
   -- * Constructors and accessors
   pollingBarNew,
+  verticalBarFromCallback,
   defaultBarConfig
   ) where
 
 import Control.Concurrent
-import qualified Control.Exception.Enclosed as E
+import Control.Exception.Enclosed ( tryAny )
 import Control.Monad ( forever )
 import Graphics.UI.Gtk
+import System.Taffybar.Widgets.Util ( backgroundLoop, drawOn )
 
 import System.Taffybar.Widgets.VerticalBar
 
-pollingBarNew :: BarConfig -> Double -> IO Double -> IO Widget
-pollingBarNew cfg pollSeconds action = do
+verticalBarFromCallback :: BarConfig -> IO Double -> IO Widget
+verticalBarFromCallback cfg action = do
   (drawArea, h) <- verticalBarNew cfg
+  drawOn drawArea $
+    backgroundLoop $ do
+      esample <- tryAny action
+      traverse (verticalBarSetPercent h) esample
 
-  _ <- on drawArea realize $ do
-    _ <- forkIO $ forever $ do
-      esample <- E.tryAny action
-      case esample of
-        Left _ -> return ()
-        Right sample -> verticalBarSetPercent h sample
-      threadDelay $ floor (pollSeconds * 1000000)
-    return ()
-
-  return drawArea
+pollingBarNew :: BarConfig -> Double -> IO Double -> IO Widget
+pollingBarNew cfg pollSeconds action =
+  verticalBarFromCallback cfg $ action <* delay
```

Comment:
I know that you didn't introduce this, but one thing that I really don't like about this approach is that it doesn't account for the fact that the action itself could take a (possibly variable) amount of time.

This means that the timing between runs might not really actually be very accurate in practice. If our action takes a non-trivial amount of time, say .5 times the interval, then it could end up getting updated far less than the user expects.

Reply:
Makes sense. Not many of the "get data" actions I have take a non-trivial amount of time, other than the one making an HTTP request to the weather service, and I'm not really worried about the accuracy of the update interval there (as long as it gets updated, period, at some point), but it does seem like something to keep in mind. I did see a few scheduling libs out there.